### PR TITLE
Rework ASF SQS provider for RegionBackend

### PR DESCRIPTION
**Pull request by viren-nadkarni**
### Summary

This PR replaces the aggregate `QueueKey` design for uniquely identifying queues with standard `RegionBackend`.
Thus enabling persistence support for SQS.

The legacy provider is removed and all integration tests updated.

### Related

Depends on https://github.com/localstack/localstack/pull/5928

### CODE REVIEW

**Comment by viren-nadkarni** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by viren-nadkarni** [comment]
Test run against pro: 
```
 =========================== short test summary info ============================
FAILED ../localstack/tests/integration/test_sns.py::TestSNSProvider::test_create_platform_endpoint_check_idempotentness
FAILED ../localstack/tests/integration/test_sns.py::TestSNSProvider::test_publish_batch_messages_from_sns_to_sqs
FAILED ../localstack/tests/integration/test_sns.py::TestSNSProvider::test_publish_batch_messages_from_fifo_topic_to_fifo_queue
FAILED ../localstack/tests/integration/test_sns.py::TestSNSProvider::test_message_to_fifo_sqs
FAILED ../localstack/tests/integration/test_sqs.py::TestSqsProvider::test_call_fifo_queue_url
= 5 failed, 892 passed, 32 skipped, 5 xfailed, 17 xpassed, 446 warnings, 10 rerun in 2474.43s (0:41:14) =
```
https://github.com/localstack/localstack/runs/6349862730?check_suite_focus=true

The failing test is correctly skipped in dd4e03a6

**Comment by thrau** [comment]
if we want to remove the old provider we also need to get #5928 fully implemented and merged

File: localstack/services/sqs/provider.py [hunk 1]
```diff
@@ -215,10 +227,9 @@ class SqsQueue:
     inflight: Set[SqsMessage]
     receipts: Dict[str, SqsMessage]
 
-    def __init__(self, key: QueueKey, attributes=None, tags=None) -> None:
-        super().__init__()
-        self._assert_queue_name(key.name)
-        self.key = key
+    def __init__(self, name: str, attributes=None, tags=None) -> None:
+        self.name = name
+        self._assert_queue_name(name)
```
Comment (thrau) [inline localstack/services/sqs/provider.py:232]: i would prefer to keep the information which region and account the queue belongs to in the resource. #5928 introduces a significant change here that quires the account id and the region to be stored in the queue. if we want to remove the `QueueKey` as a concept, then pass the region and account together with the name.

File: localstack/services/sqs/provider.py [hunk 2]
```diff
@@ -254,27 +264,16 @@ def update_last_modified(self, timestamp: int = None):
...
-    def name(self):
-        return self.key.name
-
-    @property
-    def owner(self):
-        return self.key.account_id
-
-    @property
-    def arn(self) -> str:
-        return f"arn:aws:sqs:{self.key.region}:{self.key.account_id}:{self.key.name}"
```
Comment (thrau) [inline localstack/services/sqs/provider.py:267]: i think this should stay, i don't see the value in removing it. seeing the arn of the queue in the log messages is very useful.

File: localstack/services/sqs/provider.py [hunk 3]
```diff
@@ -254,27 +264,16 @@ def update_last_modified(self, timestamp: int = None):
...
         host_url = context.request.host_url
         if config.SQS_PORT_EXTERNAL:
             host_url = external_service_url("sqs")
+
         return "{host}/{account_id}/{name}".format(
             host=host_url.rstrip("/"),
-            account_id=self.key.account_id,
-            name=self.key.name,
+            account_id=context.account_id,
+            name=self.name,
```
Comment (thrau) [inline localstack/services/sqs/provider.py:276]: this is where we need the account and region information from the queue in #5928

File: localstack/services/sqs/provider.py [hunk 4]
```diff
@@ -310,14 +309,16 @@ def remove(self, receipt_handle: str):
                 LOG.debug(
                     "no in-flight message found for receipt handle %s in queue %s",
                     receipt_handle,
-                    self.arn,
+                    self.name,
```
Comment (thrau) [inline localstack/services/sqs/provider.py:312]: as mentioned, i think we should keep the arn attribute of the resource.

File: localstack/services/sqs/provider.py [hunk 5]
```diff
@@ -719,34 +719,40 @@ def create_queue(
...
         # Special Case TODO: why is an emtpy policy passed at all? same in set_queue_attributes
         if attributes and attributes.get(QueueAttributeName.Policy) == "":
             del attributes[QueueAttributeName.Policy]
 
-        if k in self.queues:
+        if attributes is None:
+            attributes = {}
+        else:
+            attributes = attributes.copy()
+        attributes[QueueAttributeName.QueueArn] = build_arn(context, queue_name)
```
Comment (thrau) [inline localstack/services/sqs/provider.py:730]: i don't see why the QueueArn can't remain an intrinsic property of the queue

**Review by thrau — CHANGES_REQUESTED** [review]
Thanks for unifying the provider @viren-nadkarni, I do have some concerns about the changes made to the SqsQueue resource, and I would prefer to keep the region and account id explicitly in the resource.

Other than that the changes look good!

File: localstack/services/sqs/provider.py (hunk 1 above)
Comment (viren-nadkarni) [reply localstack/services/sqs/provider.py]: Good idea, made `account_id` and `region` attributes

File: localstack/services/sqs/provider.py (hunk 2 above)
Comment (viren-nadkarni) [reply localstack/services/sqs/provider.py]: Added this back now that we have this info in the queue itself

**Review by thrau — APPROVED** [review]
:rocket:

**Comment by thrau** [comment]
@dfangl i just removed the explicit test run in itest-sqs-provider, since we removed the old provider and made ASF the default

**Review by dfangl — APPROVED** [review]
CI config part :+1:

**Comment by thrau** [comment]
i was a bit over-eager approving the PR. i hadn't realized that we completely removed the old provider, but we do need a deprecation path for it. so i re-added the legacy provider so people who may rely on it can enable it with `PROVIDER_OVERRIDE_SQS=legacy`.

i also refactored a few more tests (including multi-region) and made the `create_queue` idempotent to avoid #5938

**Review by thrau — APPROVED** [review]
the scope has changed a bit now to actually making the ASF provider the default provider. which i think is great and from my side we're good to merge :-)

/cc @baermat

**Review by alexrashed — APPROVED** [review]
Great set of changes! I'm glad to see that the ASF implementation of SQS is becoming the default! 🥳

**Pull request merged**